packages/teleterm/src/ui/StatusBar/ShareFeedback/ShareFeedback.test.tsx

```diff
@@ -0,0 +1,101 @@
+import React from 'react';
+import { fireEvent, render } from 'design/utils/testing';
+import { ShareFeedback } from './ShareFeedback';
+import { MockAppContextProvider } from 'teleterm/ui/fixtures/MockAppContextProvider';
+import { ClustersService } from 'teleterm/ui/services/clusters';
+import AppContext from 'teleterm/ui/appContext';
+import { WorkspacesService } from 'teleterm/ui/services/workspacesService';
+import { Cluster } from 'teleterm/services/tshd/v1/cluster_pb';
+
+jest.mock('teleterm/ui/services/clusters');
+jest.mock('teleterm/ui/services/workspacesService');
+
+function getTestSetup() {
+  const clustersService = new ClustersService(undefined, undefined);
+  const workspacesService = new WorkspacesService(
+    undefined,
+    undefined,
+    undefined,
+    undefined
+  );
+
+  jest
+    .spyOn(ClustersService.prototype, 'findCluster')
+    .mockImplementation(clusterUri => {
+      if (clusterUri === '/clusters/localhost')
+        return {
+          loggedInUser: { name: 'alice' },
+        } as Cluster.AsObject;
+
+      if (clusterUri === '/clusters/production') {
+        return {
+          loggedInUser: {
+            name: 'bob@prod.com',
+          },
+        } as Cluster.AsObject;
+      }
+    });
+
+  jest
+    .spyOn(WorkspacesService.prototype, 'getRootClusterUri')
+    .mockReturnValue('/clusters/localhost');
+
+  // @ts-expect-error - using mocks
+  const appContext: AppContext = {
+    clustersService,
+    workspacesService,
+  };
```

Comment:
Did you consider using `MockAppContext()` to get rid of `@ts-expect-error`?

I used it in tests for `retryWithRelogin`: I call `new MockAppContext()` first and then add spies for individual functions on services.

Not the ideal solution, but it keeps typechecking intact by not using `@ts-expect-error`. TDD purists would be rather unhappy there because we mock objects that we don't own: services are instantiated within AppContext (are "owned" by AppContext) and not passed from above ("owned" by us), but well… 🙊

I think ideally we'd do something like

```typescript
const AppContextMock = AppContext as jest.MockedClass<
  typeof AppContext
>;
```

but then we'd have to mock those services anyway due to how AppContext is structured, so for now I opted to just do it as described above.

---

In general, I think [`@ts-expect-error` should be used for tests](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-9.html#ts-ignore-or-ts-expect-error) where you write type signatures so that certain code is impossible to compile (because it results in type errors) and then you can use `@ts-expect-error` to write tests for that code. But I don't think it should be used as an escape hatch for situations where the way you mock stuff doesn't play well with TypeScript. ;)

Reply:
Thanks for the advices, looks better now :)

Reply:
Nice, yeah I agree!

The conditionals you left in the mocked implementations, like

```typescript
if (clusterUri === '/clusters/localhost')
```

I don't think they're bad, it's just not something I've seen being used that often. What I'd usually see in other projects is that you'd mock a function and then verify that that function was called with correct arguments.

Reply:
Good point, I replaced these conditionals with a function call check.